servers/src/main/java/tachyon/master/file/FileSystemMaster.java

```diff
@@ -1783,6 +1806,166 @@ private void setStateFromEntry(SetStateEntry entry) throws FileDoesNotExistExcep
+    synchronized (mInodeTree) {
+      long opTimeMs = System.currentTimeMillis();
+      Inode targetInode = mInodeTree.getInodeByPath(path);
+      if (recursive && targetInode.isDirectory()) {
+        List<Inode> inodeChildren =
+            mInodeTree.getInodeChildrenRecursive((InodeDirectory) targetInode);
+        for (Inode inode : inodeChildren) {
+          setAclInternal(inode, opTimeMs, user, null, (short) -1);
+        }
+      }
+      setAclInternal(targetInode, opTimeMs, user, null, (short) -1);
+    }
+    return true;
+  }
+
+  /**
+   * Sets the group of the path. Only a super user or the path owner can change.
+   *
+   * @param path to be set group on
+   * @param group to be set
+   * @param recursive indicates whether to set all the children recursively when path is a dir
+   * @return true if set successfully, false otherwise
+   * @throws TachyonException
+   */
+  private boolean setGroup(TachyonURI path, String group, boolean recursive) throws
+      TachyonException {
+    synchronized (mInodeTree) {
+      checkOwner(path);
+      long opTimeMs = System.currentTimeMillis();
+      Inode targetInode = mInodeTree.getInodeByPath(path);
+      if (recursive && targetInode.isDirectory()) {
+        List<Inode> inodeChildren =
+            mInodeTree.getInodeChildrenRecursive((InodeDirectory) targetInode);
+        for (Inode inode : inodeChildren) {
+          checkOwner(mInodeTree.getPath(inode));
+        }
+        for (Inode inode : inodeChildren) {
+          setAclInternal(inode, opTimeMs, null, group, (short) -1);
+        }
+      }
+      setAclInternal(targetInode, opTimeMs, null, group, (short) -1);
+    }
+    return true;
+  }
+
+  /**
+   * Sets the permission of the path. Only a super user or the path owner can change.
+   *
+   * @param path to be set permission on
+   * @param permission to be set
+   * @param recursive indicates whether to set all the children recursively when path is a dir
+   * @return true if set successfully, false otherwise
+   * @throws TachyonException
+   */
+  private boolean setPermission(TachyonURI path, short permission, boolean recursive) throws
+      TachyonException {
+    synchronized (mInodeTree) {
+      checkOwner(path);
+      long opTimeMs = System.currentTimeMillis();
+      Inode targetInode = mInodeTree.getInodeByPath(path);
+      if (recursive && targetInode.isDirectory()) {
+        List<Inode> inodeChildren =
+            mInodeTree.getInodeChildrenRecursive((InodeDirectory) targetInode);
+        for (Inode inode : inodeChildren) {
+          checkOwner(mInodeTree.getPath(inode));
+        }
+        for (Inode inode : inodeChildren) {
+          setAclInternal(inode, opTimeMs, null, null, permission);
+        }
+      }
+      setAclInternal(targetInode, opTimeMs, null, null, permission);
+    }
+    return true;
+  }
+
+  private void setAclInternal(Inode inode, long opTimeMs, String user, String group,
```

Comment:
it seems we have some requirement to the caller of this method, e.g., you can only specify null to user or group to indicate not editing them, or set permission to -1 to indicate not setting it. Please make those explicit in the javadoc of this method, because they are non-trivial.